backend/inspirehep/pidstore/providers/texkey.py

```diff
@@ -0,0 +1,135 @@
+        name = name.split()[0].lower()
+        name = "".join(filter(lambda x: x in set(string.ascii_lowercase), name))
+        return name.capitalize()
+
+    @classmethod
+    def get_texkey_first_part(cls, data):
+        if "authors" in data and len(data["authors"]) < 10:
+            return cls.sanitize(get_last_name(data["authors"][0]))
+        elif "collaborations" in data:
+            return cls.sanitize(data["collaborations"][0]["value"])
+        elif "corporate_author" in data:
+            return cls.sanitize(data["corporate_author"][0])
+        elif "proceedings" in data["document_type"]:
+            return cls.sanitize("Proceedings")
+        elif "authors" in data:
+            return cls.sanitize(get_last_name(data["authors"][0]))
+        else:
+            return None
+
+    @classmethod
+    def get_texkey_second_part(cls, data):
+        date = get_literature_earliest_date(data)
+        if date is not None:
+            return date.split("-", 1)[0]
+        return None
+
+    @classmethod
+    def get_texkey_with_random_part(cls, texkey):
+        retry_count = current_app.config.get("PIDSTORE_TEXKEY_MAX_RETRY_COUNT", 5)
+        size = current_app.config.get("PIDSTORE_TEXKEY_RANDOM_PART_SIZE", 3)
+        for _ in range(retry_count):
+            random_part = "".join(
+                random.choices(string.ascii_lowercase + string.digits, k=size)
+            )
+            pid_value = f"{texkey}{random_part}"
+            if not cls.texkey_exists(pid_value):
+                return pid_value
+        return None
+
+    @classmethod
+    def get_texkey(cls, texkey):
+        texkey = (
+            PersistentIdentifier.query.with_for_update()
+            .filter(PersistentIdentifier.pid_value == texkey)
+            .filter(PersistentIdentifier.pid_type == "texkey")
+            .one_or_none()
+        )
+        return texkey
+
+    @classmethod
+    def texkey_exists(cls, texkey):
+        if cls.get_texkey(texkey):
+            return True
+        return False
+
+    @classmethod
+    def is_texkey_assigned_correctly(cls, texkey, uuid):
+        texkey = cls.get_texkey(texkey)
+        if (
+            texkey
+            and texkey.object_uuid == uuid
+            and texkey.status == PIDStatus.REGISTERED
+        ):
+            return True
+        elif texkey:
+            raise PIDAlreadyExistsError(texkey.pid_type, texkey.pid_value)
+        return False
+
+    @classmethod
+    def create(
+        cls, object_type=None, object_uuid=None, data=None, pid_value=None, **kwargs
+    ):
+        first_part = cls.get_texkey_first_part(data)
+        second_part = cls.get_texkey_second_part(data)
+        if first_part and second_part:
+            new_texkey = f"{first_part}:{second_part}"
+            pid_value = pid_value or get_value(data, "texkeys[0]")
+            if pid_value and pid_value.startswith(new_texkey):
+                if cls.is_texkey_assigned_correctly(pid_value, object_uuid):
```

Comment:
if it's not assigned correctly, then we'll not a create a `texkey` for this record?

Reply:
No. If it's assigned correctly then we are doing nothing.  if it's not assigned correctly then we are minting a new TexKey.